control/tap-ctl-ipc.c

```diff
@@ -74,13 +75,17 @@ tap_ctl_read_raw(int fd, void *buf, size_t size, struct timeval *timeout)
 			if (ret <= 0)
 				break;
 			offset += ret;
-		} else
+		} else {
+			/* 0 return - timed out */
+			err = ETIMEDOUT;
 			break;
+		}
+
 	}
 
 	if (offset != size) {
 		EPRINTF("failure reading data %zd/%zd\n", offset, size);
-		return -EIO;
+		return err ? - err : -EIO;
```

Comment:
Shouldn't this return the actual errno instead of EIO if there is one? 'err' could be set to 'errno' after 'select' and 'read', leaving 'EIO' just for the 'EOF encountered too soon' error.